### PR TITLE
perf(bitmap): optimize `bitmap::BitmapOnesIter` with bit operation

### PR DESCRIPTION
**Pull request by Honeta**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

as title

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

benchmark (unit: ns/iter)

| case                                     | baseline | current |    diff    |
|:----------------------------------------:|:--------:|:-------:|:----------:|
| bench_foldable_count_agg_vis_rate_0_05   |   55,067 |   5,522 |  -89.95 %  |
| bench_foldable_count_agg_vis_rate_0_25   |   65,652 |  24,512 |  -62.67 %  |
| bench_foldable_count_agg_vis_rate_0_5    |   86,693 |  48,290 |  -44.26 %  |
| bench_foldable_count_agg_vis_rate_0_75   |  107,777 |  74,902 |  -30.44 %  |
| bench_foldable_count_agg_without_vis     |   71,379 |  67,872 |   -4.91 %  |
| bench_foldable_max_agg_vis_rate_0_05     |   64,263 |   7,413 |  -88.47 %  |
| bench_foldable_max_agg_vis_rate_0_25     |   69,987 |  34,240 |  -51.05 %  |
| bench_foldable_max_agg_vis_rate_0_5      |   88,501 |  72,387 |  -18.22 %  |
| bench_foldable_max_agg_vis_rate_0_75     |  107,600 | 106,838 |   -0.71 %  |
| bench_foldable_max_agg_without_vis       |  116,738 |  91,278 |  -21.78 %  |
| bench_foldable_min_agg_vis_rate_0_05     |   61,119 |   7,412 |  -87.89 %  |
| bench_foldable_min_agg_vis_rate_0_25     |   81,623 |  34,118 |  -58.24 %  |
| bench_foldable_min_agg_vis_rate_0_5      |  113,487 |  72,500 |  -36.11 %  |
| bench_foldable_min_agg_vis_rate_0_75     |  122,170 | 107,285 |  -12.15 %  |
| bench_foldable_min_agg_without_vis       |   94,773 |  90,580 |   -4.42 %  |
| bench_foldable_sum_agg_vis_rate_0_05     |   63,961 |  15,735 |  -75.38 %  |
| bench_foldable_sum_agg_vis_rate_0_25     |  117,623 |  76,921 |  -34.67 %  |
| bench_foldable_sum_agg_vis_rate_0_5      |  214,199 | 154,027 |  -28.08 %  |
| bench_foldable_sum_agg_vis_rate_0_75     |  278,191 | 231,509 |  -16.80 %  |
| bench_foldable_sum_agg_without_vis       |  293,225 | 282,541 |   -3.64 %  |

### CODE REVIEW

**Comment by st1page** [comment]
>  bench_foldable_min_agg_without_vis | 229,185 | 257,951 | +12.56%

The "without_vis" is not expected to be changed. please use the same environment to rerun the test. I guess there are some performance fluctuations.

File: src/common/src/buffer/bitmap.rs [hunk 1]
```diff
@@ -340,9 +340,18 @@ impl Bitmap {
...
+                bitmap: self,
+                cur_idx: 0,
+                cur_bits: Some(self.bits[0]),
+            }
+        } else {
+            BitmapOnesIter {
+                bitmap: self,
+                cur_idx: 0,
+                cur_bits: None,
+            }
```
Comment (kwannoel) [inline src/common/src/buffer/bitmap.rs:354]:
```suggestion
        let cur_bits = if self.num_bits > 0 { Some(self.bits[0]) } else { None };
        BitmapOnesIter {
            bitmap: self,
            cur_idx: 0,
            cur_bits,
        }
```

File: src/common/src/buffer/bitmap.rs [hunk 2]
```diff
@@ -631,18 +640,31 @@ unsafe impl TrustedLen for BitmapIter<'_> {}
...
+        }
+        match self.cur_bits {
+            Some(bits) => {
+                let low_bit = bits & bits.wrapping_neg();
+                let low_bit_idx = bits.trailing_zeros();
+                self.cur_bits = Some(bits ^ low_bit);
+                Some(self.cur_idx * BITS + low_bit_idx as usize)
+            }
+            _ => None,
         }
```
Comment (st1page) [inline src/common/src/buffer/bitmap.rs:664]:
```suggestion
        self.cur_bits.map(|bits| {
            let low_bit = bits & bits.wrapping_neg();
            let low_bit_idx = bits.trailing_zeros();
            self.cur_bits = Some(bits ^ low_bit);
            Some(self.cur_idx * BITS + low_bit_idx as usize)
        })
```

File: src/common/src/buffer/bitmap.rs [hunk 3]
```diff
@@ -631,18 +637,31 @@ unsafe impl TrustedLen for BitmapIter<'_> {}
...
 impl<'a> iter::Iterator for BitmapOnesIter<'a> {
     type Item = usize;
 
     fn next(&mut self) -> Option<Self::Item> {
-        let next_idx = self.bitmap.next_set_bit(self.idx);
-        if let Some(idx) = next_idx {
-            self.idx = idx + 1;
+        while self.cur_bits == Some(0) {
+            self.cur_idx += 1;
+            self.cur_bits = if self.cur_idx == (self.bitmap.len() + BITS - 1) / BITS {
```
Comment (kwannoel) [inline src/common/src/buffer/bitmap.rs:650]: This part doesn't seem very readable. Can you explain what `(self.bitmap.len() + BITS - 1) / BITS` does in a doc comment?

Comment (kwannoel) [reply src/common/src/buffer/bitmap.rs:650]: Or refactor it into a variable

Comment (st1page) [inline src/common/src/buffer/bitmap.rs:650]: A little tricky but I think it is ok
```suggestion
            self.cur_bits = if self.cur_idx >= self.bits.len() {
```

Comment (st1page) [reply src/common/src/buffer/bitmap.rs:650]: in fact, `Bitmap::vec_len()`

**Comment by st1page** [comment]
> bench_foldable_sum_agg_without_vis | 691,241 | 602,814 | -12.79%

I am not sure if EC2 do something bad :innocent:

**Review by TennyZhuang — APPROVED** [review]

**Comment by lmatz** [comment]
Does the benchmark numbers in the top thread come from using deterministic SEED or a random one?

**Comment by Honeta** [comment]
> Does the benchmark numbers in the top thread come from using deterministic SEED or a random one?

Currently random. #9280 will solve the problem probably.

File: src/common/src/buffer/bitmap.rs [hunk 4]
```diff
@@ -631,18 +637,28 @@ unsafe impl TrustedLen for BitmapIter<'_> {}
 
 pub struct BitmapOnesIter<'a> {
     bitmap: &'a Bitmap,
-    idx: usize,
+    cur_idx: usize,
+    cur_bits: Option<usize>,
```
Comment (kwannoel) [inline src/common/src/buffer/bitmap.rs:641]: I suspect we can save a usize here.

```rs
size_of Option<usize>: 16
// change to usize: 8
```

None case is terminal. Hence `cur_idx` is no longer needed when that happens.
We can encode `None` case as `cur_idx == usize::MAX`, as I don't think we will have bitmap of such a large size.
Then `cur_bits` will just be `usize` and we use `cur_idx` as the discriminant for `None` case.

Don't need to implement in this PR as it adds complexity, or possibly ever, just noting it for my own reference.
Only if Heap profile shows bitmap iter takes an unreasonable amount of memory, we should optimize it (not a priority now).

**Review by st1page — APPROVED** [review]
LGTM

**Comment by Honeta** [comment]
> Does the benchmark numbers in the top thread come from using deterministic SEED or a random one?

Updated.

**Pull request merged**